### PR TITLE
[WIP] Feature Powerlimiter

### PR DESCRIPTION
**Pull request by berni2288**
Todo: Write list of features & changes here.

### CODE REVIEW

File: src/MqttSettings.cpp [hunk 1]
```diff
@@ -35,6 +47,14 @@ void MqttSettingsClass::onMqttConnect(bool sessionPresent)
     for (const auto& cb : _mqttSubscribeParser.get_callbacks()) {
         mqttClient->subscribe(cb.topic.c_str(), cb.qos);
     }
+
```
Comment (helgeerbe) [inline src/MqttSettings.cpp:50]: This is not used and callbacks point to 0. I would remove all changes to this file.

File: src/PowerLimiter.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+    _lastLoop = millis();
+
+    std::shared_ptr<InverterAbstract> inverter = Hoymiles.getInverterByPos(0);
+
+    if (inverter == nullptr || !inverter->isReachable()) {
+        return;
+    }
+
+    float dcVoltage = inverter->Statistics()->getChannelFieldValue(CH1, FLD_UDC);
```
Comment (helgeerbe) [inline src/PowerLimiter.cpp:86]: must be 
float dcVoltage = inverter->Statistics()->getChannelFieldValue(TYPE_DC, CH1, FLD_UDC);

File: src/PowerLimiter.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    Hoymiles.getMessageOutput()->printf("[PowerLimiterClass::loop] victronChargePower: %d\n",
+        static_cast<int>(victronChargePower));
+
+    if (millis() - _lastPowerMeterUpdate < (30 * 1000)) {
+        Hoymiles.getMessageOutput()->printf("[PowerLimiterClass::loop] dcVoltage: %f config.PowerLimiter_VoltageStartThreshold: %f config.PowerLimiter_VoltageStopThreshold: %f inverter->isProducing(): %d\n",
+            dcVoltage, config.PowerLimiter_VoltageStartThreshold, config.PowerLimiter_VoltageStopThreshold, inverter->isProducing());
+    }
+
+    if (inverter->isProducing()) {
+        float acPower = inverter->Statistics()->getChannelFieldValue(CH0, FLD_PAC);
```
Comment (helgeerbe) [inline src/PowerLimiter.cpp:103]: Must be
float acPower = inverter->Statistics()->getChannelFieldValue(TYPE_AC, CH0, FLD_PAC);

File: src/MqttSettings.cpp (hunk 1 above)
Comment (berni2288) [reply src/MqttSettings.cpp:50]: This is not coming from my branch. Maybe a rebase problem? I don't have access to my IDE at the moment. When I've time, I'll do one more merge or rebase and make sure that everything is correct.

**Comment by Adminius** [comment]
May be it make sense to accept my PR #102 before this will be integrated into development branch?
than I can do my changes in development (there is not much work to do any more)
I'll try to make good working version today and give my go, OK?

**Comment by berni2288** [comment]
You can just change the target branch of your PR to development and I'll also continue working in the powerlimiter branch after the first merge, so there is no need to hurry for you :)

**Comment by helgeerbe** [comment]
@berni2288  I merged your power_limiter branch to develop. It's working fine. So I propose you create a new feature branch from development for further power_limiter changes.

**Comment by berni2288** [comment]
Thanks! I can create a branch for each pull request from now on.

**Comment by berni2288** [comment]
Closing this PR since the branch already got merged to "development" manually.

**Pull request closed without merging**